bundles/core/src/main/java/com/adobe/cq/wcm/core/components/util/ComponentUtils.java

```diff
@@ -35,139 +36,167 @@
+	 * @return The ID property value if set, empty optional if not.
+	 */
+	private static Optional<String> getPropertyId(@NotNull final Resource resource) {
+		return Optional.ofNullable(resource.getValueMap().get(Component.PN_ID, String.class))
+				.filter(StringUtils::isNotEmpty).map(StringUtils::trim).map(StringUtils::normalizeSpace)
+				.map(val -> StringUtils.replace(val, " ", ID_SEPARATOR));
+	}
+
+	/**
+	 * Returns an auto generated component ID.
+	 *
+	 * The ID is the first {@value ComponentUtils#ID_HASH_LENGTH} characters of an
+	 * SHA-1 hexadecimal hash of the component path, prefixed with the component
+	 * name. Example: title-810f3af321
+	 *
+	 * If the component is referenced, the path is taken to be a concatenation of
+	 * the component path, with the path of the first parent context resource that
+	 * exists on the page or in the template. This ensures the ID is unique if the
+	 * same component is referenced multiple times on the same page or template.
+	 *
+	 * Collision --------- c = expected collisions c ~= (i^2)/(2m) - where i is the
+	 * number of items and m is the number of possibilities for each item. m = 16^n
+	 * - for a hexadecimal string, where n is the number of characters.
+	 *
+	 * For i = 1000 components with the same name, and n = 10 characters:
+	 *
+	 * c ~= (1000^2)/(2*(16^10)) c ~= 0.00000045
+	 *
+	 * @return the auto generated component ID
+	 */
+	@NotNull
+	private static String generateId(@NotNull final Resource resource, @Nullable final Page currentPage,
+			@Nullable final ComponentContext componentContext) {
+		String resourceType = resource.getResourceType();
+		String prefix = StringUtils.substringAfterLast(resourceType, "/");
+		String path = resource.getPath();
+		if (currentPage != null && componentContext != null) {
+			PageManager pageManager = currentPage.getPageManager();
+			Page resourcePage = pageManager.getContainingPage(resource);
+			Template template = currentPage.getTemplate();
+			boolean inCurrentPage = (resourcePage != null
+					&& StringUtils.equals(resourcePage.getPath(), currentPage.getPath()));
+			boolean inTemplate = (template != null && path.startsWith(template.getPath()));
+			if (!inCurrentPage && !inTemplate) {
+				ComponentContext parentContext = componentContext.getParent();
+				while (parentContext != null) {
+					Resource parentContextResource = parentContext.getResource();
+					if (parentContextResource != null) {
+						Page parentContextPage = pageManager.getContainingPage(parentContextResource);
+						inCurrentPage = (parentContextPage != null
+								&& StringUtils.equals(parentContextPage.getPath(), currentPage.getPath()));
+						inTemplate = (template != null
+								&& parentContextResource.getPath().startsWith(template.getPath()));
+						if (inCurrentPage || inTemplate) {
+							path = parentContextResource.getPath().concat(resource.getPath());
+							break;
+						}
+					}
+					parentContext = parentContext.getParent();
+				}
```

Comment:
It looks like your IDE auto applied line length restrictions. 
I'm not sure that anything in this file actually needs to be changed, but I could be missing something since the diff covers most of the file.

Reply:
Ah, i see what's happened, you have changed the files to tab indentation instead of space indentation. 
Does this project have an actual style guide?
I think everything else is always space indentation.

Reply:
@ky940819 - Not that I am aware of. But if you want me to change it to original format, I can do that. Please let me know. Thanks !